### PR TITLE
feat: workflow status listener - RK-19190 

### PR DESCRIPTION
**Pull request by gosharo**
Signed-off-by: Gosha <gosha@rookout.com>

### Description

Please provide a brief description of the changes made in this pull request.

### Related Issue(s)

If this pull request addresses or relates to any open issues, please mention them here using the syntax `Fixes #<issue_number>` or `Resolves #<issue_number>`.

### Checklist

Before submitting this pull request, please ensure that you have completed the following tasks:

- [x] Reviewed the [Contributing Guidelines](../docs/CONTRIBUTING.md) for the Piper project.
- [x] Ensured that your changes follow the [coding guidelines and style](../docs/CONTRIBUTING.md#coding-guidelines) of the project.
- [x] Run the tests locally and made sure they pass.
- [x] Updated the relevant documentation, if applicable.

### Testing Instructions

Please provide clear instructions on how to test and verify the changes made in this pull request.

### Additional Information

Add any additional information or context that would be helpful in understanding and reviewing this pull request.

### CODE REVIEW

File: cmd/piper/piper.go [hunk 1]
```diff
@@ -51,10 +55,34 @@ func main() {
 		panic(err)
 	}
 
-	//err = common.Git.UnsetWebhook()
-	//if err != nil {
-	//	panic(err)
-	//}
+	ctx := context.Background()
```
Comment (qadiludmer) [inline cmd/piper/piper.go:58]: I would split the code to 2 files


event_handlers.go
```go
func WatchWorkflowEvents(workflowClient) {
 // create the watch and start listening to events and also handling them
}
```

piper.go (this is the main)
```go
go WatchWorkflowEvents(globalClients.Workflows)
```

This way piper.go only deals with initializing stuff and not doing any actual logic

**Review by qadiludmer — COMMENTED** [review]
see comment

File: pkg/event_handler/event_handler.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"k8s.io/apimachinery/pkg/watch"
+	"log"
+	"strconv"
+)
+
+type eventHandlerImpl struct{}
+
+func (eh *eventHandlerImpl) handle(workflowChan <-chan watch.Event) {
```
Comment (qadiludmer) [inline pkg/event_handler/event_handler.go:13]: handle should get the actual event, not the channel

File: pkg/event_handler/main.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+func Start(cfg *conf.GlobalConfig, clients *clients.Clients) {
+	ctx := context.Background()
+	watcher, err := clients.Workflows.Watch(&ctx)
+	if err != nil {
+		log.Panicf("Failed to watch workflow error:%s", err)
+	}
+
+	handlerImpl := &eventHandlerImpl{}
+	go func() {
+		handlerImpl.handle(watcher.ResultChan())
```
Comment (qadiludmer) [inline pkg/event_handler/main.go:19]: here you should wait for channel data and once new data arrive pass it to the handler

File: pkg/event_handler/event_handler.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	log.Printf("event handler started")
+	for event := range workflowChan {
+		wf, ok := event.Object.(*v1alpha1.Workflow)
```
Comment (qadiludmer) [inline pkg/event_handler/event_handler.go:16]: type checking should be done on the `Start()` method

**Review by qadiludmer — COMMENTED** [review]
see comments

**Comment by gosharo** [comment]
@qadiludmer followed you comments, and eventually erased the handler. think that I will bring it back

**Review by qadiludmer — APPROVED** [review]

**Pull request merged**